reactioncommerce/reaction
# feat: GQL discount & discount codes plugins

Resolves #5752 
Resolves: #5758 
Impact: **major**  
Type: **feature**

## Issue

Need GQL queries and mutations for discount codes

## Solution

- Create queries and mutations for discount codes.
- Discounts Code are live in the `Discounts` collection
- Add integration tests for Discount Codes

## Breaking changes

## Testing
1. Add discount
1. Query discounts
1. Update discounts
1. Query discounts again to verify
1. Delete discounts
1. Query again

GraphQL for your queries / mutations
```graphql
query PrimaryShopId {
	primaryShopId
}

query DiscountCodes($shopId: ID!) {
  discountCodes(shopId: $shopId) {
    nodes {
      _id
      label
      code
      shop {
        _id
      }
    }
  }
}

mutation CreateDiscountCode($shopId: ID!) {
  createDiscountCode(input: {
    shopId: $shopId,
    discountCode: {
      code: "LUNCH ALL"
      description: "Monies Off of all lunch"
      discountMethod: code
      calculation: {
        method: discount
      }
			conditions: {
        accountLimit: 1
				order: {
          min: 0.00
          max: 400.00
          startDate: "2019-11-14T18:30:03.658Z"
          endDate: "2021-01-01T08:00:00.000Z"
        }
        redemptionLimit: 0
        audience: ["customer"]
        permissions: ["guest", "anonymous"]
        products: ["product-id"]
        tags: ["tag-id"]
        enabled: true
      }
      transactions: [{
        cartId: "cart-id"
        userId: "user-id"
        appliedAt: "2019-11-18T18:30:03.658Z"
      }]
    }
  }) {
    discountCode {
      _id
      shop {
        _id
      }
      code
      label
      description
      discountMethod
      discount
      transactions {
        cartId
        userId
        appliedAt
      }
      calculation {
        method
      }
      conditions {
        accountLimit
        audience
        enabled
        permissions
        redemptionLimit
				order {
          min
          max
          startDate
          endDate
        }
        products
        tags
      }
    }
  }
}

mutation UpdateDiscountCode($discountCodeId: ID!, $shopId: ID!) {
  updateDiscountCode(input: {
    discountCodeId: $discountCodeId,
    shopId: $shopId,
    discountCode: {
      code: "LUNCH ALL"
      description: "Monies Off of all lunch"
      discountMethod: code
      calculation: {
        method: discount
      }
			conditions: {
        accountLimit: 1
				order: {
          min: 0.00
          max: 400.00
          startDate: "2019-11-14T18:30:03.658Z"
          endDate: "2021-01-01T08:00:00.000Z"
        }
        redemptionLimit: 0
        audience: ["customer"]
        permissions: ["guest", "anonymous"]
        products: ["product-id"]
        tags: ["tag-id"]
        enabled: true
      }
      transactions: [{
        cartId: "cart-id"
        userId: "user-id"
        appliedAt: "2019-11-18T18:30:03.658Z"
      }]
    }
  }) {
    discountCode {
      _id
      shop {
        _id
      }
      code
      label
      description
      discountMethod
      discount
      transactions {
        cartId
        userId
        appliedAt
      }
      calculation {
        method
      }
      conditions {
        accountLimit
        audience
        enabled
        permissions
        redemptionLimit
				order {
          min
          max
          startDate
          endDate
        }
        products
        tags
      }
    }
  }
}

mutation DeleteDiscountCode($discountCodeId: ID!, $shopId: ID!) {
  deleteDiscountCode(input: {
    discountCodeId: $discountCodeId,
    shopId: $shopId
  }) {
    discountCode {
      _id
      shop {
        _id
      }
      code
      label
      description
      discountMethod
      discount
      transactions {
        cartId
        userId
        appliedAt
      }
      calculation {
        method
      }
      conditions {
        accountLimit
        audience
        enabled
        permissions
        redemptionLimit
				order {
          min
          max
          startDate
          endDate
        }
        products
        tags
      }
    }
  }
}
```
